### PR TITLE
Add CircleCI

### LINKED ISSUE #440: Investigate/switch CI
**Issue summary**

With countless issues with Github actions and now builds running out of storage space, either a workaround needs to be added so this doesn't fail or switch over to another CI. 

Discussion indicated CircleCI would be a good switch, but open to discussion.

**Other information and links**

### PR DESCRIPTION
**Pull request by GregTheGreek**
**Summary of changes**
Changes introduced in this pull request:
- Adds circle ci
- Uses a circleci machine instead of docker as intended originally.

**Reference issue to close (if applicable)**

Closes #440 

**Other information and links**

I suggest rolling this out slowly, not making circle builds mandatory *yet*, let a few prs go through to iron out any potential bugs before migrating.

### CODE REVIEW

**Comment by GregTheGreek** [comment]
@priom I agree

One thing I've done is actually use a cache on the cargo builds so that we don't need to re-build everything

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,146 @@
+version: 2.1
+
+executors:
+  test-executor:
+    machine:
+      image: ubuntu-1604:202004-01
```
Comment (austinabell) [inline .circleci/config.yml:6]: Are you sure ubuntu 16 is required?

File: .circleci/config.yml [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+    description: Save cargo package cache for subsequent jobs
+    steps:
+      - save_cache:
+          key: cargo-package-cache-{{ .Environment.CIRCLE_SHA1 }}
+          # paths are relative to /home/circleci/project/
+          paths:
+            - ../.cargo/git
+            - ../.cargo/registry
+            - ../.cargo/.package-cache
+            # - ../../../usr/local/include
```
Comment (austinabell) [inline .circleci/config.yml:77]:
```suggestion
```
? Or is there a convenience need to keep

Comment (GregTheGreek) [reply .circleci/config.yml:77]: Proroc is stored in there

File: .circleci/config.yml (hunk 1 above)
Comment (ec2) [reply .circleci/config.yml:6]: Nope. Its just the lastest VM image CircleCI has

File: .circleci/config.yml (hunk 2 above)
Comment (ec2) [reply .circleci/config.yml:77]: im gunna remove it. Dont think its wise to store all the headers lol

Comment (austinabell) [reply .circleci/config.yml]: the headers are in the submodule? What's wrong with caching protoc?

**Pull request merged**